### PR TITLE
Mast machine mapping

### PR DESCRIPTION
**Pull request by jmcclena**
This pull request adds initial machine mapping for MAST for constructing EFIT inputs

### CODE REVIEW

File: omas/machine_mappings/mast.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    Loads MAST tokamak poloidal field coil hardware geometry
+
+    :param ods: ODS instance
+    """
+    from omfit_classes.omfit_efund import OMFITmhdin, OMFITnstxMHD
+
+    mhdin = get_support_file(OMFITmhdin, nstx_filenames('mhdin', pulse))
+    mhdin.to_omas(ods, update=['oh', 'pf_active'])
+
+    signals = get_support_file(OMFITnstxMHD, nstx_filenames('signals', pulse))
```
Comment (smithsp) [inline omas/machine_mappings/mast.py:44]: `nstx_filenames` in `mast.py`?

File: omas/machine_mappings/mast.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+            else:
+                cid = icoil_signals[c_pf]['mds_name'].strip('\\').replace("~", " ")
+        for e in ods[f'pf_active.coil'][c]['element']:
+            # if 'OH' in ods[f'pf_active.coil'][c]['name']:
+            #    ename = oh_signals[c_oh]['mds_name_resolved'].strip('\\') + f'_element_{e}'
+            # else:
+            #    ename = icoil_signals[c_pf]['mds_name_resolved'].strip('\\') + f'_element_{e}'
+            ename = cid + f'_element_{e}'
+            eid = ename
+            #ods[f'pf_active.coil'][c]['name'] = cname
```
Comment (smithsp) [inline omas/machine_mappings/mast.py:72]: Cruft?
```suggestion
```

File: omas/machine_mappings/mast.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+            cname = icoil_signals[c_pf]['name']
+            if icoil_signals[c_pf]['mds_name'] is None:
+                cid = 'None'
+            else:
+                cid = icoil_signals[c_pf]['mds_name'].strip('\\').replace("~", " ")
+        for e in ods[f'pf_active.coil'][c]['element']:
+            # if 'OH' in ods[f'pf_active.coil'][c]['name']:
+            #    ename = oh_signals[c_oh]['mds_name_resolved'].strip('\\') + f'_element_{e}'
+            # else:
+            #    ename = icoil_signals[c_pf]['mds_name_resolved'].strip('\\') + f'_element_{e}'
```
Comment (smithsp) [inline omas/machine_mappings/mast.py:69]: Cruft or something that needs further attention?
```suggestion
```

File: omas/machine_mappings/mast.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    :param ods: ODS instance
+
+    :param pulse: shot number
+    """
+    from omfit_classes.omfit_efund import OMFITnstxMHD
+    from omfit_classes.utils_math import firFilter
+    ods1 = ODS()
+    unwrap(pf_active_hardware)(ods1, pulse)
+
+    signals = get_support_file(OMFITnstxMHD, nstx_filenames('signals', pulse))
```
Comment (smithsp) [inline omas/machine_mappings/mast.py:92]: `OMFITnstxMHD` and `nstx_filenames` in `mast.py`?

File: omas/machine_mappings/mast.py [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+        if f'magnetics.flux_loop.{channel}.flux.data' in ods:
+            data = ods[f'magnetics.flux_loop.{channel}.flux.data']
+            rel_error = data * tfl_signals[channel + 1]['rel_error']
+            abs_error = tfl_signals[channel + 1]['abs_error']
+            error = np.sqrt(rel_error**2 + abs_error**2)
+            error[np.abs(data) < tfl_signals[channel + 1]['sig_thresh']] = tfl_signals[channel + 1]['sig_thresh']
+            ods[f'magnetics.flux_loop.{channel}.flux.data_error_upper'] = error
+            # 2*pi normalization is done at this stage so that rel_error, abs_error, sig_thresh are consistent with data
+            ods[f'magnetics.flux_loop.{channel}.flux.data'] /= 2.0 * np.pi
+            ods[f'magnetics.flux_loop.{channel}.flux.data_error_upper'] /= 2.0 * np.pi
```
Comment (smithsp) [inline omas/machine_mappings/mast.py:210]: Is this the right approach @orso82 ?

**Review by smithsp — CHANGES_REQUESTED** [review]
Some odd `nstx` lingering in `mast.py`

**Comment by smithsp** [comment]
@jmcclena You should merge master in here, to get the automatic COCOS conversions of `error_upper` that @bechtt implemented.

**Comment by jmcclena** [comment]
@bpatel2107

File: omas/machine_mappings/mast.py (hunk 5 above)
Comment (jmcclena) [reply omas/machine_mappings/mast.py]: Removed in 859246a8a214f617334e02c4f4bb32ca14a1ea06

File: omas/machine_mappings/mast.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+@machine_mapping_function(__regression_arguments__, pulse=44653)
+def magnetics_hardware(ods, pulse):
+    r"""
+    Load MAST tokamak flux loops and magnetic probes hardware geometry
+
+    :param ods: ODS instance
+    """
+    from omfit_classes.omfit_efund import OMFITmhdin, OMFITnstxMHD
+
+    mhdin = get_support_file(OMFITmhdin, nstx_filenames('mhdin', pulse))
```
[the remark this answers is not in the document]
Comment (jmcclena) [reply omas/machine_mappings/mast.py]: changed nstx_filename to mast_filenames

File: omas/machine_mappings/mast.py (hunk 4 above)
Comment (jmcclena) [reply omas/machine_mappings/mast.py]: changed nstx_filenames to mast_filenames

File: omas/machine_mappings/mast.py (hunk 3 above)
Comment (jmcclena) [reply omas/machine_mappings/mast.py]: removed

**Pull request merged**